sync/src/snapshot/service.rs

```diff
@@ -55,16 +60,88 @@ impl ChainNotify for Service {
             .map(|hash| self.client.block_number(BlockId::Hash(*hash)).expect("Enacted block must exist"))
             .any(|number| number % self.period == 0);
         if is_checkpoint && best_number > self.period {
+            let client = self.client.clone();
             let root_dir = self.root_dir.clone();
             let period = self.period;
             spawn(move || {
-                let target = (best_number / period - 1) * period;
-                let path: PathBuf = [root_dir, target.to_string()].iter().collect();
-                if let Ok(_) = create_dir_all(path) {
-                    // FIXME: implement this
-                    // unimplemented!()
+                let number = (best_number / period - 1) * period;
+                let header = client.block_header(BlockId::Number(number)).expect("Snapshot target must exist");
+                let path: PathBuf = [root_dir, format!("{:x}", header.hash())].iter().collect();
+                match create_dir_all(&path) {
+                    Ok(_) => {}
+                    // FIXME: Handle file system errors
+                    Err(_) => return,
+                }
+                let db = client.database();
```

Comment:
As we had talked F2F, creating the snapshot right after the block generation can be dangerous because of reorganization.

Reply:
Current implementation doesn't take snapshot on best block.
Snapshot target must be something like `floor(best_number - snapshot_period, snapshot_period)`.